### ISSUE TITLE
Variable declaration error in `for` expression

### ISSUE THREAD

**Issue by pawwe** [hashicorp/vscode-terraform#1693]
### Extension Version

`v2.29.4`

### VS Code Version

```
Version: 1.86.1
Commit: 31c37ee8f63491495ac49e43b8544550fbae4533
Date: 2024-02-07T09:07:19.849Z
Electron: 27.2.3
ElectronBuildId: 26495564
Chromium: 118.0.5993.159
Node.js: 18.17.1
V8: 11.8.172.18-electron.0
OS: Linux x64 6.7.4-200.fc39.x86_64
```

### Operating System

`Linux fws 6.7.4-200.fc39.x86_64 #1 SMP PREEMPT_DYNAMIC Mon Feb  5 22:21:14 UTC 2024 x86_64 GNU/Linux`

### Terraform Version

`Terraform v1.7.3 on linux_amd64`

### Steps to Reproduce

Define a variable that is a list of objects. Use the variable in a `for` expression in a `for_each` argument. Then I see the error `No declaration found for "var.VAR_NAME"`. But only when used in a `for_each` argument. No error when used in a `output` value for example.

![image](https://github.com/hashicorp/vscode-terraform/assets/9119956/f7dfb3f3-f67a-4edc-b7e0-97eea6a28eae)

### Expected Behavior

No errors

### Actual Behavior

I get this error:

```json
[
  {
    "resource": "/home/fws/tf_test/main.tf",
    "owner": "HashiCorpTerraform",
    "severity": 8,
    "message": "No declaration found for \"var.foobar\"",
    "source": "Terraform",
    "startLineNumber": 12,
    "startColumn": 25,
    "endLineNumber": 12,
    "endColumn": 35
  }
]
```

### Terraform Configuration

```terraform
variable "foobar" {
  type = list(object({
    name = string
  }))
  default = [
    { name = "foo" },
    { name = "bar" },
  ]
}

resource "local_file" "foobar" {
  for_each = { for i in var.foobar : i.name => i }

  content  = each.value.name
  filename = "${path.module}/${each.value.name}.txt"
}

output "foobar" {
  value = { for i in var.foobar : i.name => i }
}
```

### Workarounds

Downgrading to `v2.29.3` or setting `terraform.validation.enableEnhancedValidation` to `false` does not trigger this error.

### Help Wanted

- [ ] I'm interested in contributing a fix myself

### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by KazuSh1geru** [comment]
I'm the same.

**Comment by dbanck** [comment]
Thanks for the report! I can reproduce the issue with the snippet you provided.

**Comment by dbanck** [comment]
The underlying issue here is, that `{ for ... }` [will result in an object](https://developer.hashicorp.com/terraform/language/expressions/for#result-types), but the `for_each` meta-argument [accepts a map or a set of strings](https://developer.hashicorp.com/terraform/language/meta-arguments/for_each#basic-syntax). Passing objects to `for_each` is still supported for backwards compatibility reasons, but should generally be avoided. 

A workaround is to wrap the expression in a `toset` call:
```hcl
resource "local_file" "foobar" {
  for_each = toset({ for i in var.foobar : i.name => i })

  content  = each.value.name
  filename = "${path.module}/${each.value.name}.txt"
}
```
This will get rid of the error and fix the completion and go to definition for `var.foobar`.

The validation error here is still worth fixing though, because it is a valid Terraform configuration. In an ideal world, we would raise a warning instead and allow you to apply a quickfix.

**Comment by scott-doyland-burrows** [comment]
It doesn't always pick up on that as an error. I have that type of code all over the place, and not all show as errors.

**Comment by dbanck** [comment]
We just released a new version [2.29.5](https://github.com/hashicorp/vscode-terraform/releases/tag/v2.29.5). This fixes the reported bug. The update should appear automatically in VS Code.

Thank you for the report @pawwe 

--- 

In case you experience any different validation related bug, please do let us know through a new issue.
